### ISSUE TITLE
error：当采用pointwise损失，并将负采样设置为4时，recall@20出现1.0的情况

### ISSUE THREAD

**Issue by sunshinelium** [wubinzzu/NeuRec#26]
您好， 我在测试这些方法时，在ml-100k数据集上，使用loo划分，99个测试负采样。训练时的负采样设置为4，采用cross_entropy损失。得到了以下结果：
![image](https://user-images.githubusercontent.com/44607300/86019333-0a8ba100-ba59-11ea-8597-459f450388f0.png)
recall@20达到了1.0，而recall@10的值却很低。由于之前遇到过这种情况，当时是由于topK排序算法在排序过程中，对于相同的值会默认选择第一个，当模型得到的预测值非常接近且目标物品在第一个时就会出现hr，recall=1.0的情况。但是我看了下这部分的代码，由于编码能力有限，我先将evaluator.uni_evaluator.py中的evaluate function中的
`candidate_items = [list(self.user_pos_test[u]) + self.user_neg_test[u] for u in batch_users]`
改成了
`candidate_items = [self.user_neg_test[u] +list(self.user_pos_test[u]) for u in batch_users]`
这在我之前碰到的情况中是有效的。但这里他没有影响。
![image](https://user-images.githubusercontent.com/44607300/86021863-2e9cb180-ba5c-11ea-84de-82f4553080b9.png)

   @@@而这个地方的test_items我认为两种情况下应该是一样的？这个是指g_item（目标物品）么？为什么当指定负测试样本时，test_items是这样的呢？我搞不太清楚。tool.argmax_top_k我也搞不太懂。非常希望得到您的解答，或者一个提示。非常感谢

**Comment by sunshinelium** [comment]
此外，在测试集负采样矩阵不为None的情况下，pointwise的训练集负采样除了应排除训练集正样本外，还应排除测试机中的负样本，避免数据泄露。
![image](https://user-images.githubusercontent.com/44607300/86081033-e36bb880-bac6-11ea-9035-b690d507d0dd.png)
而这个地方只去掉了训练集正样本。期待您的回复，非常感谢，希望有机会与您共同探讨

**Comment by ZhongchuanSun** [comment]
> 您好， 我在测试这些方法时，在ml-100k数据集上，使用loo划分，99个测试负采样。训练时的负采样设置为4，采用cross_entropy损失。得到了以下结果：
> ![image](https://user-images.githubusercontent.com/44607300/86019333-0a8ba100-ba59-11ea-8597-459f450388f0.png)
> recall@20达到了1.0，而recall@10的值却很低。由于之前遇到过这种情况，当时是由于topK排序算法在排序过程中，对于相同的值会默认选择第一个，当模型得到的预测值非常接近且目标物品在第一个时就会出现hr，recall=1.0的情况。但是我看了下这部分的代码，由于编码能力有限，我先将evaluator.uni_evaluator.py中的evaluate function中的
> `candidate_items = [list(self.user_pos_test[u]) + self.user_neg_test[u] for u in batch_users]`
> 改成了
> `candidate_items = [self.user_neg_test[u] +list(self.user_pos_test[u]) for u in batch_users]`
> 这在我之前碰到的情况中是有效的。但这里他没有影响。
> ![image](https://user-images.githubusercontent.com/44607300/86021863-2e9cb180-ba5c-11ea-84de-82f4553080b9.png)
> 
> @@@而这个地方的test_items我认为两种情况下应该是一样的？这个是指g_item（目标物品）么？为什么当指定负测试样本时，test_items是这样的呢？我搞不太清楚。tool.argmax_top_k我也搞不太懂。非常希望得到您的解答，或者一个提示。非常感谢

一、先说模型的结果, 这是我自己跑出来的结果:
![screenshot](https://user-images.githubusercontent.com/10204480/86109603-22b1fd80-baf7-11ea-95f4-9f70f0e1d295.png)
看起来还算正常, Top10的结果不算低, Top20的结果也没有达到1.0.
你这个结果是改动了某些代码之后得出的么?

二、self.eval_score_matrix这个函数参数ranking_score的第u行第i列是用户u对物品i的评分, 也就是说, 这个函数将矩阵的列下标作为了物品的id使用.
所以:
    1) 在全样本测试的时候, 评分矩阵的列数和物品个数是相等, 而且列下标和物品的id是相等的, 因此可以直接使用评分矩阵和测试集中的物品id来计算模型的推荐性能.
    2) 在基于负样本进行测试的时候, 评分矩阵的列数和物品个数是不等的, 列下标和物品的id也不存在对应相等的关系. 这种情况下, 对于某一用户, 其评分向量包含三部分: 测试物品评分, 负样本评分和填充的负无穷. 在具体的实现中, 我们将测试物品放在了最前面, 然后负样本, 最后是负无穷. 这样排序之后,相当于对物品的id做了重映射, 测试物品的id也就变成了0~len-1, len是测试物品的个数.

三、关于tool.argmax_top_k: 在计算物品TopK的排序列表的时候, 我们并不需要对所有物品排序, 只要保证得到评分最大的K物品id, 且有序就可以了. 因此我们对物品列表使用了部分排序, 这样可以将列表降低排序的时间复杂度由O(nlogn)降低到O(n). 在cpp的测试代码中, 我们使用的也是这种策略.

四、不知道你图中贴出来的实验结果, 是不是应为改变了测试物品和负样本的顺序导致的.

**Comment by ZhongchuanSun** [comment]
> 此外，在测试集负采样矩阵不为None的情况下，pointwise的训练集负采样除了应排除训练集正样本外，还应排除测试机中的负样本，避免数据泄露。
> ![image](https://user-images.githubusercontent.com/44607300/86081033-e36bb880-bac6-11ea-9035-b690d507d0dd.png)
> 而这个地方只去掉了训练集正样本。期待您的回复，非常感谢，希望有机会与您共同探讨

以下是我个人的理解, 不一定正确:
    严格意义上来说, 我们是不知道测试集中都有哪些物品的. 所以, 在对数据进行操作的时候, 是不能考虑测试集的. 在为测试阶段采负样本的时候, 如果排除了测试集中的物品, 可能会导致指标的值比现在的更高. 而且, 虽然测试阶段的负样本没有参与模型的训练, 但如果考虑排除测试集中的物品, 这可能更像是数据泄露.

**Comment by sunshinelium** [comment]
不好意思，之前好像忘记说了，我跑的是MLP.py，没有改动程序，结果还是那样，您这个结果是跑的哪个方法啊？还有关于数据泄露的问题。如果训练集的负采样不去掉测试集时，可能出现测试集中的负样例在训练集中出现的情况，这时，由于模型已经学习过这个负样本，所以会出现测试效果偏高的情况。我是这样想的，不过我看到的代码里好像也都不是这样考虑的，所以我就探讨一下，希望不要见怪。哈哈

**Comment by ZhongchuanSun** [comment]
> 不好意思，之前好像忘记说了，我跑的是MLP.py，没有改动程序，结果还是那样，您这个结果是跑的哪个方法啊？还有关于数据泄露的问题。如果训练集的负采样不去掉测试集时，可能出现测试集中的负样例在训练集中出现的情况，这时，由于模型已经学习过这个负样本，所以会出现测试效果偏高的情况。我是这样想的，不过我看到的代码里好像也都不是这样考虑的，所以我就探讨一下，希望不要见怪。哈哈

我看了下, 基于负采样测试运行MLP有点问题,会报错, 问题在MLP.py的135行,现在已经改过了. 改过之后再运行,结果是正常的.
![screenshot_1](https://user-images.githubusercontent.com/10204480/86467474-04066d80-bd68-11ea-9249-a652569cc077.png)

你在运行下, 如果还有问题, 把NeuRec.properties和MLP.properties两个配置文件的内容贴一下我看看.

关于负样本那部分, 没太明白你说的什么意思, 能再详细点么?

**Comment by ZhongchuanSun** [comment]
> 此外，在测试集负采样矩阵不为None的情况下，pointwise的训练集负采样除了应排除训练集正样本外，还应排除测试机中的负样本，避免数据泄露。
> ![image](https://user-images.githubusercontent.com/44607300/86081033-e36bb880-bac6-11ea-9035-b690d507d0dd.png)
> 而这个地方只去掉了训练集正样本。期待您的回复，非常感谢，希望有机会与您共同探讨

这个是哪个版本的代码, 我刚看了下最新的代码, 好像和这个不太一样

**Comment by sunshinelium** [comment]
NeuRec.properties
`[default]
recommender=MLP

config_dir=./conf
gpu_id=0
gpu_mem=0.5

data.input.path=dataset
data.input.dataset= ml-100k

data.column.format=UIRT

data.convert.separator='\t'

user_min=0
item_min=0

splitter=loo

ratio=0.8
by_time=False

metric=["Precision", "Recall", "NDCG", "MAP", "MRR"]

topk=[10, 20]

group_view=None
rec.evaluate.neg=99
test_batch_size=1024
num_thread=8`

MLP.properties
`[hyperparameters]
epochs=100
batch_size=256
layers=[64,32,16]
reg_mlp=0.0
learning_rate=0.001
learner=adam
is_pairwise=False
num_neg=4

loss_function=cross_entropy
verbose=1
init_method=normal
stddev=0.01`
您修改的那个地方我之前看到改过来了，结果还是那个结果。😂
还有负采样的那个地方是data/sampler.py中的，是跟最新版本一致的，我是觉得如果训练集负采样只排除正样本时，存在与测试集负采样重叠的可能性的话，可能会让模型在测试部分作弊，所以结果会偏高一点儿，但目前看到的代码都是这样做的，所以这个其实算我的一个疑问😄

**Comment by ZhongchuanSun** [comment]
![screenshot](https://user-images.githubusercontent.com/10204480/86874083-3fab8980-c112-11ea-96ce-3274721b11b3.png)
这是我跑出来的结果, 在Windows和Ubuntu系统都试了, 实验结果都试这个样子的.

关于训练集的采样, 在训练的时候, 我们所了解的只是仅限于训练集, 我们不知道除此之外任何信息, 因此在训练的时候只排除了训练集中的正样本, 测试集中的正样本都没有排除.

**Comment by sunshinelium** [comment]
非常感谢，我自己再找找问题所在吧，谢谢了，😄

**Comment by sunshinelium** [comment]
我又看了一下，我怀疑NeuRec/model/general_recommender/MLP.py的这个地方您可能没看到
![image](https://user-images.githubusercontent.com/44607300/87009135-54ae1880-c1f7-11ea-8e56-c6fbedae7c5d.png)
如果只是改了超参数那块，结果其实还是负采样为1的结果

**Comment by ZhongchuanSun** [comment]
@sunshinelium 

恩恩,这个之前没有注意到, 按照这个修改之后, 指标计算上的确出现了你说的问题.
出现问题的具体原因:
    1) MLP的激活函数使用的是Relu, 这使得用户对物品的评分为非负, 会出现0的情况.
    2) 当负样本为3或4及更多时, 会导致MLP中用户的对物品的评分中会有很多的0, 最大多数评分都为0, 这个可以通过调试查看.
    3) 基于负采样做测试的时候, 我们的做法是将正样本的id重映射到0, 也就是用户评分向量的第一个元素.
    4) 在上述情况同时发生时, cpp里面的排序函数'std::partial_sort_copy' 会将正样本放在ranking list 的topk前面的位置, 也就导致了评测结果的异常.

上述问题已经做了修正, 并已提交 https://github.com/wubinzzu/NeuRec/commit/908b18b4c8ac260bc1c69717a7ec71550cfdce98, 你可以再试下最新版本的代码.

最后, 非常感谢你对NeuRec的支持和提交的bug!
